src/Monotonic.cpp

```diff
@@ -30,26 +31,204 @@ using std::string;
 
 namespace {
 
-class MonotonicVisitor : public IRVisitor {
+bool is_constant(const ConstantInterval &a) {
+    return a.is_single_point(0);
+}
+
+bool may_be_negative(const ConstantInterval &a) {
+    return !a.has_lower_bound() || a.min < 0;
+}
+
+bool may_be_positive(const ConstantInterval &a) {
+    return !a.has_upper_bound() || a.max > 0;
+}
+
+bool is_monotonic_increasing(const ConstantInterval &a) {
+    return !may_be_negative(a);
+}
+
+bool is_monotonic_decreasing(const ConstantInterval &a) {
+    return !may_be_positive(a);
+}
+
+ConstantInterval to_interval(Monotonic m) {
+    switch (m) {
+    case Monotonic::Constant:
+        return ConstantInterval::single_point(0);
+    case Monotonic::Increasing:
+        return ConstantInterval::bounded_below(0);
+    case Monotonic::Decreasing:
+        return ConstantInterval::bounded_above(0);
+    case Monotonic::Unknown:
+        return ConstantInterval::everything();
+    }
+    return ConstantInterval::everything();
+}
+
+Monotonic to_monotonic(const ConstantInterval &x) {
+    if (is_constant(x)) {
+        return Monotonic::Constant;
+    } else if (is_monotonic_increasing(x)) {
+        return Monotonic::Increasing;
+    } else if (is_monotonic_decreasing(x)) {
+        return Monotonic::Decreasing;
+    } else {
+        return Monotonic::Unknown;
+    }
+}
+
+ConstantInterval unify(const ConstantInterval &a, const ConstantInterval &b) {
+    return ConstantInterval::make_union(a, b);
+}
+
+ConstantInterval unify(const ConstantInterval &a, int64_t b) {
+    ConstantInterval result;
+    result.include(b);
+    return result;
+}
+
+// Helpers for doing arithmetic on ConstantIntervals that avoid generating
+// expressions of pos_inf/neg_inf.
+ConstantInterval add(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+    result.min_defined = a.has_lower_bound() && b.has_lower_bound();
+    result.max_defined = a.has_upper_bound() && b.has_upper_bound();
+    if (result.has_lower_bound()) {
+        result.min = a.min + b.min;
+    }
+    if (result.has_upper_bound()) {
+        result.max = a.max + b.max;
+    }
+    return result;
+}
+
+ConstantInterval add(const ConstantInterval &a, int64_t b) {
+    return add(a, ConstantInterval(b, b));
+}
+
+ConstantInterval negate(const ConstantInterval &r) {
+    ConstantInterval result;
```

Comment:
If `ConstantInterval`s should maintain the invariant that `min < max` regardless of `min_defined`/`max_defined`, then this function doesn't necessarily follow that - it could be simplified by replacing:
```
result.min = r.has_upper_bound() ? -r.max : 0;
```
with:
```
result.min = -r.max;
```
And likewise for result.max. Nothing about the ConstantInterval would change, but it would maintain the `min < max` invariant (not sure if this invariant is intended/useful, but it does remove ternary operators that are unnecessary

Reply:
I don't think this is safe, because negation can overflow. 

I think we only need min < max when min_defined && max_defined. 

I think the combination of these is why there are so many ternarys where they seem possibly redundant. I think many of them are unnecessary except to avoid possible ubsan issues due to overflow.

Reply:
Ah, that makes sense. Should we be concerned about overflow in other cases? Because I believe that `multiply(ConstantInterval, int64_t)` is passed a `uint64_t` at some point, which could be an issue (I assumed that you were assuming no overflow)

Reply:
I was more worried about introducing overflow from earlier implementations of is_monotonic. These ternaries don't protect against generating new overflow (just like basically all integer arithmetic).

I decided to just fix the need for a signed/unsigned case anyways, that was always super annoying.